### DISCUSSION TITLE
Difficulties using RandZoom transform

### DISCUSSION THREAD

**Discussion by LaiaHumbertVidan** [Project-MONAI/MONAI#2653] [General]
Hi, 

I am able to train my network with RandRotation transform applied to the data. However, when I also want to apply the RandZoom transform, I get the following error:

RuntimeError: applying transform <monai.transforms.compose.Compose object at 0x000001BFB6257128>

Here is how I am coding this:

train_transforms = Compose(
        [AddChannel(), 
         RandRotate(range_x=0.1, range_y=0.1,range_z=0.1, prob=0.6, padding_mode='zeros'), 
         RandZoom(prob=0.1, min_zoom=0.0, max_zoom=0.0, mode='linear', padding_mode='linear_ramp', 
         keep_size=True),
         ToTensor() 
         ]) 

train_ds = monai.data.ImageDataset(image_files=ID_train, labels=y_train, transform=train_transforms, image_only=True) 

batch_size = 8 # explored 5 to 30 and 20 seems to give best results with densenet121
train_loader = monai.data.DataLoader(train_ds, batch_size=batch_size, pin_memory=torch.cuda.is_available())

im = monai.utils.first(train_loader) 
print(im[0].shape) 
 

Many thanks in advance!

**Comment by Nic-Ma** [comment]
Hi @LaiaHumbertVidan ,

Thanks for your interest and experiments with MONAI.
Could you please help paste the full error log for debug? The log line you pasted is not the key message.

Thanks in advance.

**Comment by LaiaHumbertVidan** [comment]
Hi, sorry. Here is the full error message:

=== Transform input info -- RandZoom ===
Data statistics:
Type: <class 'numpy.ndarray'>
Shape: (1, 64, 64, 53)
Value range: (0.0, 0.8540586233139038)
Default upsampling behavior when mode=linear is changed to align_corners=False since 0.4.0. Please specify align_corners=True if the old behavior is desired. See the documentation of nn.Upsample for details.
Traceback (most recent call last):

  File "<ipython-input-19-6b5f68107be4>", line 1, in <module>
    im = monai.utils.first(train_loader) # get the first item from the iterable

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\monai\utils\misc.py", line 71, in first
    for i in iterable:

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\torch\utils\data\dataloader.py", line 345, in __next__
    data = self._next_data()

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\torch\utils\data\dataloader.py", line 385, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\torch\utils\data\_utils\fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\torch\utils\data\_utils\fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\monai\data\image_dataset.py", line 111, in __getitem__
    img = apply_transform(self.transform, img)

  File "C:\Users\lh18\AppData\Local\Continuum\anaconda2\envs\lhv\lib\site-packages\monai\transforms\transform.py", line 71, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e

RuntimeError: applying transform <monai.transforms.compose.Compose object at 0x000002A9DCF5B208>
________________________________

**Reply by Nic-Ma** [reply]
Hi @LaiaHumbertVidan ,

Do you have any more error log after "RuntimeError: applying transform <monai.transforms.compose.Compose"?
I didn't see any error in the log related to `Randzoom` transform.

Thanks.

**Comment by LaiaHumbertVidan** [comment]
Hi,

This error only comes up if I try to include the RandZoom transform, it doesn't happen if I run the code with the other transforms only.
There is no other error log than what I pasted in the previous message, I'm afraid.

I have tried to get more info about the error by adding this: CUDA_LAUNCH_BLOCKING="1" and running it on cpu device (it has worked before in other cases) but did not get any further.


________________________________

**Reply by Nic-Ma** [reply]
Hi @LaiaHumbertVidan ,

Maybe it's because the arg values in your `RandZoom` is not expected? `min_zoom=0.0, max_zoom=0.0`
Can you try again default values?

Thanks.

**Comment by LaiaHumbertVidan** [comment]
Hi,

Many thanks for your suggestion. I've tried min_zoom=0.9, max_zoom=1.1 as in the documentation example but it doesn't seem to make it work as I still get the same problem. Are RandZoom and RandRotate compatible and can they be used at the same time?


________________________________

**Reply by Nic-Ma** [reply]
Hi @LaiaHumbertVidan ,

I tried to develop a program to reproduce your issue with synthetic data:
```py
print(f"generating synthetic data to {tempdir} (this may take a while)")
for i in range(40):
    im, seg = create_test_image_3d(128, 128, 128, num_seg_classes=1)

    n = nib.Nifti1Image(im, np.eye(4))
    nib.save(n, os.path.join(tempdir, f"im{i:d}.nii.gz"))

    n = nib.Nifti1Image(seg, np.eye(4))
    nib.save(n, os.path.join(tempdir, f"seg{i:d}.nii.gz"))

images = sorted(glob(os.path.join(tempdir, "im*.nii.gz")))
segs = sorted(glob(os.path.join(tempdir, "seg*.nii.gz")))

# define transforms for image and segmentation
train_imtrans = Compose(
    [
        AddChannel(),
        RandRotate(range_x=0.1, range_y=0.1,range_z=0.1, prob=1.0, padding_mode='zeros'),
        RandZoom(prob=1.0, min_zoom=0.0, max_zoom=0.0, mode='linear', padding_mode='linear_ramp', keep_size=True),
        EnsureType(),
    ]
)

# define image dataset, data loader
check_ds = ImageDataset(images, segs, transform=train_imtrans, image_only=True)
check_loader = DataLoader(check_ds, batch_size=10, num_workers=2, pin_memory=torch.cuda.is_available())
im, seg = monai.utils.misc.first(check_loader)
print(im.shape, seg.shape)
```
And it printed out the real error:
```
File "/workspace/data/medical/MONAI/monai/transforms/spatial/array.py", line 574, in __call__
  zoomed = torch.nn.functional.interpolate(  # type: ignore
File "/opt/conda/lib/python3.8/site-packages/torch/nn/functional.py", line 3726, in interpolate
  raise NotImplementedError("Got 5D input, but linear mode needs 3D input")
```
Seems you should not use the `linear` as the interpolation mode in `RandZoom` transform?
When I changed to `mode='trilinear', padding_mode='constant'`, no issue showed anymore.

Thanks.

**Comment by LaiaHumbertVidan** [comment]
Hi,

It worked! Thank you so much for your help with this, very much appreciated.
________________________________

**Reply by Nic-Ma** [reply]
No problem, please feel free to raise a ticket if you face any other issues.

Thanks.
